### DISCUSSION TITLE
Methods for large feature datasets

### DISCUSSION THREAD

**Discussion by Nova-Scotia** [R-ArcGIS/arcgislayers#68] [Q&A] [answered]
Hi @JosiahParry. I've tested the "overwrite" capability of {arcgislayers} and it works nicely on a small test dataset - horay! When I tried it on a bigger dataset though (43,000 records, 15 variables), I got an error:

```
Error in `httr2::req_perform()`:
! HTTP 413 Payload Too Large.
Backtrace:
 1. arcgislayers::add_features(flayer, newdata_sp)
 2. httr2::req_perform(req)
```

I'm also running into issues as I'd like my field Aliases to be different than the field names - not sure how to do this when writing the layer using {arcgisbinding} - so far I was just editing them in Pro then sharing it to AGOL. 

It looks like the [truncate and append python script](https://community.esri.com/t5/arcgis-online-documents/overwrite-arcgis-online-feature-service-using/ta-p/904457) manages larger data by chunking it up then sending it piece by piece.

**Comment by JosiahParry** [comment]
Thank you! There are two problems here:

1. Pagination of requests to make sure that the feature size is not too big
2. Identifying which names to match fields based on.

## Request Pagination

The issue that you're running into is that you're trying to send _too_ much data over http. arcgislayers uses `{httr2}` which uses curl to send request. There's only so much data that can be sent at a time. There is an existing issue for this: https://github.com/R-ArcGIS/arcgislayers/issues/11

@jthomasmock provided a [nice solution for chunking a data frame](https://gist.github.com/jthomasmock/0215f9d2787aa3ef6149ef48fd4d038b) that we can take advantage of here in the interim. 

The idea is that we can use the `chunk_df()` function to partition a data frame into chunks of `n` rows. Then we can add each of those chunks to a feature service iteratively using a for loop or lapply or purrr or what ever you'd like. 

```r
library(dplyr)
library(arcgislayers)

# https://gist.github.com/jthomasmock/0215f9d2787aa3ef6149ef48fd4d038b
# vector operation
chunk_vec <- function(chunk_size){
  ceiling(seq_along(row_number())/chunk_size)
}

# operate on dataframe
chunk_df <- function(.data, chunk_size) {
  
  mutate(.data, chunk = chunk_vec(chunk_size)) %>%
    group_by(chunk) %>%
    group_split()
  
}

# create a sample spatial data frame 
n <- 45000
sdf <- tibble(
  fid = seq_len(n),
  x = rnorm(n),
  y = runif(n)
) |> 
  sf::st_as_sf(coords = c("y", "x"))


# split based on a chunk size
sdf_splits <- chunk_df(sdf, 5000)

# iteratively add the features 
for (i in seq_along(sdf_splits)) {
  add_features(x, sdf_splits[[i]])
}
```

## Matching fields

Right now `add_features()` and `update_features()` requires that the column names either match the field alias of the field name verbatim. Note that there is the `match_on` argument which, by default, matches on the alias. If you want to match on the field name set `match_on = "field"`. 

Is there an alternate that you would prefer? I think it would be feasible to provide a named vector or a named list to act as a lookup table between the data frame and the featuer service. But in essence, we _need_ to have a way to determine which columns match to which fields in the feature service.

**Reply by JosiahParry** [reply, marked as answer]
@Nova-Scotia a new PR https://github.com/R-ArcGIS/arcgislayers/pull/72 has been added to arcgislayers to automatically chunk the request and sends the requests in parallel. I tested with a table feature service with 45k features and it took ~5 secs. You can specify the size of the chunks using `chunk_size` argument which is set to `2000` by default. If you have a lot of columns or complex geometries consider reducing the size of that argument.

**Reply by Nova-Scotia** [reply]
Oh, wow. This is so excellent - what a gift to come into work to find out there is active development on the file size, and an answer to my question about field names! I'll test and see what happens - will get back to you asap!

**Comment by Nova-Scotia** [comment]
Yes! Just added 14,393 features to my large feature dataset and it worked like a charm 😭😭😭 These are happy tears. Thank you so very much! Matching on name worked a treat.

**Reply by JosiahParry** [reply]
yay! Amazing!

**Comment by JosiahParry** [comment]
@Nova-Scotia better late than never! We can now truncate feature layers with the `truncate_layer()` funciton: https://github.com/R-ArcGIS/arcgislayers/pull/107
